complex/ft_msg.c

```diff
@@ -124,12 +143,24 @@ static int ft_post_send(void)
 		msg.addr = ft_tx_ctrl.addr;
 		msg.context = NULL;
 		msg.data = 0;
-		ret = fi_sendmsg(ft_tx_ctrl.ep, &msg, 0);
+		ft_send_retry(ret, fi_sendmsg, ft_tx_ctrl.ep, &msg, 0);
 		ft_next_iov_cnt(&ft_tx_ctrl, fabric_info->tx_attr->iov_limit);
+		ft_tx_ctrl.credits--;
+		break;
+	case FT_FUNC_INJECT:
+		ft_send_retry(ret, fi_inject, ft_tx_ctrl.ep, ft_tx_ctrl.buf,
+				ft_tx_ctrl.msg_size, ft_tx_ctrl.addr);
+		break;
+	case FT_FUNC_INJECTDATA:
+		ft_send_retry(ret, fi_injectdata, ft_tx_ctrl.ep, ft_tx_ctrl.buf,
+				ft_tx_ctrl.msg_size, ft_tx_ctrl.remote_cq_data,
+				ft_tx_ctrl.addr);
 		break;
 	default:
-		ret = fi_send(ft_tx_ctrl.ep, ft_tx_ctrl.buf, ft_tx_ctrl.msg_size,
-				ft_tx_ctrl.memdesc, ft_tx_ctrl.addr, NULL);
+		ft_send_retry(ret, fi_send, ft_tx_ctrl.ep, ft_tx_ctrl.buf,
+				ft_tx_ctrl.msg_size, ft_tx_ctrl.memdesc,
+				ft_tx_ctrl.addr, NULL);
+		ft_tx_ctrl.credits--;
```

Comment:
We're losing the return values from the send calls.

I'd also like to see if we can avoid using macros.  It's much easier to read/maintain code that is directly calling a function, versus using macro magic to construct the function call.  I'm just tossing out ideas now, but conceptually the code is doing:

prepare_send - e.g. ft_format_iov
do_send - fi_send*
post_send_processing - e.g. ft_next_iov_cnt, credits--

The EAGAIN loop is only want to work over the 'do_send' portion.  Maybe the code would be cleaner if it were restructured to this format?  The prepare section is dependent on the class_function.  The do send section is dependent on the class_function and test option (use credits or EAGAIN), as is the post processing.

I think that change can follow this patch as a possible restructuring, in case the resulting changes end up making the code worse.  For this specific patch, the only change I would make is to set ret, to avoid losing any errors.